src/page/slot.rs

```diff
@@ -266,21 +214,25 @@ impl<T, C: cfg::Config> Slot<T, C> {
         // will have to be removed when those references are dropped.
         if refs.value > 0 {
             return true;
+        } else {
+            false
         }
-
-        // Otherwise, we can remove the slot now!
-        test_println!("-> remove deferred; can remove now");
-        let removed = self.remove_value(curr_gen, offset, free).is_some();
-        removed
     }
 
-    #[inline]
-    pub(super) fn remove_value<F: FreeList<C>>(
+    /// Mutates this slot.
+    ///
+    /// This method spins until no references to this slot are left, and calls the mutator
+    fn mutate_storage_with_confirmation<F, M, R>(
         &self,
         gen: Generation<C>,
         offset: usize,
         free: &F,
-    ) -> Option<T> {
+        mutator: M,
+    ) -> Option<R>
+    where
+        F: FreeList<C>,
+        M: FnOnce(*mut T) -> R,
```

Comment:
IMO, this method is responsible for ensuring that it's safe to mutate the value, which it does by spinning until the ref count is 0, so i don't think the function that's passed in needs to be unsafe. we could change this to
```suggestion
        M: FnOnce(&mut T) -> R,
```
and then below, where we call it, we could say 
```rust
let value = self.item.with_mut(|item| mutator(unsafe { &mut *item }));
```
that way, this method is responsible for ensuring the mutation is safe, rather than the caller.